### ISSUE TITLE
migrate rancher owned cluster to another rancher deployment

### ISSUE THREAD

**Issue by aemneina** [rancher/rancher#16471]
**What kind of request is this (question/bug/enhancement/feature request):**
feature request

**Use Case:**
Ability to migrate clusters to be managed by another rancher deployment.
This should maintain the same ability to scale out custom clusters.

**Comment by dsargent3220** [comment]
This is also VERY useful in the rare case where there is a disaster and the Rancher configuration is lost and a user needs to rebuild Rancher from the ground up.  In this case there is no way to import the previous clusters that Rancher had created.  The clusters would need to be recreated, which while doable, is very painful.

**Comment by cjellick** [comment]
@aemneina  are you requesting this for hosted clusters, custom, imported, or all of the above?
Also, what do you mean by "This should maintain the same ability to scale out custom clusters." I dont quite understand what you mean by that requirement

**Comment by goffinf** [comment]
I raised a support request (with ZenDesk) where the response was that THIS ticket was sited by Rancher Support as the potential fix (see: https://support.rancher.com/hc/en-us/requests/4208)

To be clear, our requirement is to be able to import Application Clusters to a specific Rancher management cluster through automation.

Automation is the critical factor since we don't allow use of the Rancher management UI for effecting change in our Production environment. Any changes are only possible using automated release pipelines which include change manager approval and creation/update of change audit records.

I won't replay all of the support ticket response but in a nutshell it is covered by 

> ... there isn't a way to import a cluster to another rancher installation at this time however we have a GitHub issue open for this. https://github.com/rancher/rancher/issues/16471

**Comment by aemneina** [comment]
@cjellick Ideally all of the above on cluster type. Understandable if we'd have to prioritize some over others.

What I meant with: "This should maintain the same ability to scale out custom clusters." is the case I have a cluster created by rancher using node templates. I have the ability there to scale the node pools, this most likely will be lost migrating the managed cluster to another rancher server instance.

Not sure if we can meet @goffinf 's requirement by cleaning up the cluster of rancher cruft, then create a new cluster for import on another managed rancher. I think the only piece missing in that case @goffinf is that cleanup step, this should be straight forward to automate. I'll sync with @patrick0057 on this.

**Comment by goffinf** [comment]
@cjellick thanks. If some of the ‘cruft’ you are referring are the hosts that are still registered to Rancher but marked as unavailable because they actually no longer exist but Rancher doesn’t know that, we do run periodic de-register script that tidies those up. We did the same on v1.x and again on 2.x to deal with situations where you cycle a node out of a cluster and replace it with another and something goes wrong. We have a mandatory control that means we must replace our whole estate of nodes (of all types) frequently (min once a month) and of course we must be prepared for both DR and any security vulnerability that might occur at any time that requires us to burn everything.

Or perhaps you are referring to some other cruft (no shortage of that around here ;-) )

**Comment by dsargent3220** [comment]
Our R&D department is looking at using Rancher / RKE to build clusters as well.  They have a bit of a unique need to be able to auto-deploy a cluster onto know hardware (using RKE) and then import and manage that cluster with Rancher at a later time, or even move the cluster from one Rancher to another.

**Comment by livenson** [comment]
We have a similar requirement - to be able to move K8S clusters between supported and self-supported environments.

**Comment by stieler-it** [comment]
We created an RKE cluster and now want to move it to a new Rancher 2 with HA setup. Seems also be waiting for a solution to this issue?

**Comment by stieler-it** [comment]
Is there a way to manually achieve it?

**Comment by walkafwalka** [comment]
Where are the configurations for Rancher stored now?

**Comment by intonet** [comment]
I'am waiting for this feature. I want to migrate clusters to a closer ( new dc/ the same region), fresh rancher.

**Comment by wzrdtales** [comment]
technically it should be just a matter of exporting all the custom crds + objects from the local cluster to the new one and referencing those right afterwards.

**Comment by savealive** [comment]
We'd love this feature too. We've tried to move rancher crds using Velero but didn't work.

**Comment by adacaccia** [comment]
Importing a Rancher-owned RKE cluster to another Rancher instance can be viewed as a feature, but I insist on it as it is crucial in any disaster scenario.
As an example, one of my Rancher installations died yesterday because of that cruft cert-manager, so the procedure highlighted [here](https://rancher.com/docs/rancher/v2.x/en/upgrades/upgrades/ha/) (more precisely, the "OPTION B: REINSTALLING RANCHER AND CERT-MANAGER") didn't work this time!
Now I'm left with a running pre-production RKE cluster, which I'm currently managing via kubectl with an updated kubeconfig file.
Importing that cluster in a fresh Rancher installation would just ease any upcoming management activity, which is exactly the Rancher mantra itself!
A robust K8s orchestrator shouldn't have any problems in moving clusters from an installation to another!

**Comment by aekiratli** [comment]
There is one case that is really frustrating, prod and test clusters running in single node docker rancher, I need to to seperate them but I cant find a way to migrate test cluster to another rancher instance.

**Comment by aemneina** [comment]
2.5 should have a backup/restore (think migration via restores) feature, I suspect that would meet the need. I have yet to try it myself but encourage anyone with a test cluster to kick the tires.

**Comment by mscbpi** [comment]
Maybe with the register feature that allows managing a cluster "as if" it was created with rancher ? Then maybe it's capable of importing exsting k8s cluster that would have been created with another rancher instance.

**Comment by travis-sobeck** [comment]
@deniseschannon @cjellick 
I'm a bit unclear, is there an intent to build a feature to all migrating an existing Ranching cluster from one Rancher instance to another?
Thanks

**Comment by HugoKuo** [comment]
Is cluster migration between Rancher server implemented?

**Comment by kucerarichard** [comment]
Is there a way to just delete the existing agent in a rancher-managed cluster and then re-import it into another rancher server by just that import agent deployment from the new cluster?

**Comment by Gallardo994** [comment]
Huge bump. I would like to move from single-node-installation to HA in a different datacentre. Currently this doesn't look like an option, but might still be doable, e.g.:
1. Export all kubeconfigs
2. Install new Rancher in a new location
3. Create clusters as "Imported"
4. Execute kubectl commands to import clusters
At least I've managed to successfully perform full Rancher destruction and reinstallation without any old data on HA cluster, but it was on Digitalocean DOKS rather than custom RKE cluster.

**Comment by dnoland1** [comment]
@Gallardo994 If moving all clusters over, you can do a Rancher backup and restore. This will even work if moving from different types of clusters (RKE, K3S, AKS, EKS, GKE, etc). See https://rancher.com/docs/rancher/v2.x/en/backups/v2.5/

**Comment by binhbo12456** [comment]
any news ?

**Comment by BBQigniter** [comment]
please, add this this feature would have been awesome in my case :(

**Comment by mattmattox** [comment]
Could we use the Rancher backup operator to collect all the different resources IE cluster object, machines, projects, rolebindings, etc for a cluster? Then import it into another Rancher IE kubectl apply all that YAML?

**Comment by harridu** [comment]
This is **so much** needed. Please implement.

In fact, what I need is some way to do regular backups all necessary information about a cluster needed to import it into another Rancher instance later. Something to get production up and running asap.

**Comment by krishofmans** [comment]
@harridu yes, this is all well documented: https://rancher.com/docs/rancher/v2.5/en/backups/

**Comment by mattmattox** [comment]
> @harridu yes, this is all well documented: https://rancher.com/docs/rancher/v2.5/en/backups/

This documentation covers migrating the Rancher server (upstream) between clusters. This feature request is to move a downstream cluster to another Rancher deployment. For example, I could deploy a downstream Cluster "A" in Rancher-A.example.net then later down the road I might want to move that cluster to Rancher-B.example.net.

This is currently there is no supported way to do this in Rancher.

**Comment by jaskirat8** [comment]
+1 on this as I am just in this situation where I whole machine and DR having rancher is lost forever due to some mess up in infra management with no scope of recovery. 

Since RKE is running fine as 3 node cluster and even kubeconfig is also available for controlpane access.

But all management capabilities are lost and this feature, where spinning up new rancher and importing this cluster there will help me regain management access.

Even if there is a manual way to do this, please guide me.

**Comment by laurie-kepford** [comment]
+1 for me too! I need this ability to manage disaster recovery.  It would also give me the ability to "promote" a test cluster that is attached to a test rancher to be a production cluster on a production rancher.

**Comment by FredNass** [comment]
Same here. We'd like to move a cluster from the production Rancher instance to the test Rancher instance. Ideally, it would be nice to have an "Unregister" button next to the "Import Existing" button in the cluster management view. Hope you guys can add this feature in next Rancher versions.

**Comment by stex79** [comment]
Clearly this is an important feature that would help in a lot of different scenarios.
It does not look like difficult to implement, are there any strategical/political reasons to not do it?

**Comment by plusiv** [comment]
Still waiting for this feature!!!

**Comment by laurie-kepford** [comment]
I think all that thats really needed is a reliable and safe way to Remove Rancher from a cluster.  After that you could just import it where you want it to be.  I have seen some instructions on removing rancher, but none that are officially supported by rancher.  Wait!  I found this: https://ranchermanager.docs.rancher.com/faq/rancher-is-no-longer-needed#what-if-i-dont-want-rancher-anymore.  Would that be the solution?

**Comment by laurie-kepford** [comment]
Answering my own question.  The method above removes rancher "and anything installed by rancher".  This would not be desirable in cases where rancher was use to install apps like maybe datadog or neuvector.  For me, it might works since I install everything with helm.  But its a little scary.  HOWEVER, in my case the section titled "What if I don't want my registered cluster managed by Rancher?" https://ranchermanager.docs.rancher.com/faq/rancher-is-no-longer-needed#what-if-i-dont-want-rancher-anymore would probably be my best solution.

In my environment I create the management cluster with eskctl, install rancher on it.  Then create my downstream clusters and import them.  I haven't test it, but seems like just simply deleting it from one Rancher Manager and importing it into another one, would work.

Obviously this would not work for rancher-created clusters because the described action would delete the entire cluster.

**Comment by laurie-kepford** [comment]
i tested the above and it works fine.  It removed all the rancher components from the cluster and left it intact.

**Comment by voarsh2** [comment]
My Rancher install died/stopped working correctly, span up a new instance, did the unrecommended "import existing Rancher managed cluster" - showed up, but I can't upgrade Kubernetes version or anything like that.... plus it hasn't really worked either.

**Comment by plusiv** [comment]
> My Rancher install died/stopped working correctly, span up a new instance, did the unrecommended "import existing Rancher managed cluster" - showed up, but I can't upgrade Kubernetes version or anything like that.... plus it hasn't really worked either.

One of the biggest problems with importing an existing cluster into Rancher is that you lose some features that you had when Rancher provisioned the cluster.

**Comment by JannikZed** [comment]
This would make a lot of Ops way more easy! Our Rancher instance is already very old and it seems like some things get broken with many upgrades and migration tasks. Some resources are left behind and our rancher is no longer working as expected - I was spending already several days, trying to "repair" things, but it just stays broken and does not upgrade downstream clusters any longer. I would just like to throw it all away and start with a new Rancher ..

**Comment by philipp1992** [comment]
we require this feature as well, so move a lot of legacy clusters away from an unupgradable single instance rancher server. Any ETA?

**Comment by clp01** [comment]
Any update on this one?

**Comment by egallis31** [comment]
Haven't tested this yet but could the latest rancher-backup chart be used to backup the resources required and redeploy to a "migrated" instance? 

https://ranchermanager.docs.rancher.com/how-to-guides/new-user-guides/backup-restore-and-disaster-recovery/back-up-rancher

**Comment by jattind** [comment]
Importing orphaned downstream clusters into a new Rancher instance seems pretty essential, but I am surprised that this requirement has not yet been implemented, even after so many requests for this feature and such long time. I would have to reconsider using Rancher as an enterprise solution for my Kubernetes strategy. Please provide an update or let us know if you don't intend to ever implement this. Thanks.

**Comment by FredNass** [comment]
I totally agree. Not having the ability to recover an orphaned downstream cluster in Rancher's Management introduces a major operational risk that enterprises should not have to take.

**Comment by Avneetdabas** [comment]
Is there an update on this. We are also thinking of using rancher as our enterprise kubernetes solution but need this feature to be in place before we take this decision.

**Comment by jostmart** [comment]
same here

**Comment by txj-xyz** [comment]
Following this for updates, this feature would be required for us.

**Comment by ronnyaa** [comment]
following for updates, have used rancher a while and been happy. But this is a major roadblock 
Is there any way to do this manually?

**Comment by bananflugan** [comment]
It has now been many years since this feature request, which is a very important one. Customers need to be able to migrate their downstream clusters to new Rancher instances. Please look into this request and sort it out.

**Comment by kucerarichard** [comment]
this isn't there yet?!    Was just jumping onto Rancher Prime with the hope I'd be able to spin up rke2 clusters,  and have them be somewhat independent should the Rancher server have update failure or other such meltdowns.   Was hoping to lean on Rancher and not have to consume the Ansible playbooks for rke2...   The devs really lean on Rancher too,  having a reliable web-based hub gets rid of Ansible (tower) and also gets rid of any build directories full of machine specs and cluster notes,  no terraform, it'd all be in the UI.  alright I'm over it,  moving on.

**Comment by erdeerdeerde** [comment]
Any update?

**Comment by JoshuaFurman** [comment]
No movement on this issue after 6 years?? This is an essential feature for DR.

**Comment by maxisam** [comment]
> Actually it is not a viable option. It removes all the projects, and therefore all the namespaces.

@laurie-kepford Could you explain what happened?  We are in the same boat. Did you try to do this step  "What if I don't want my registered cluster managed by Rancher?" Are those clusters created by Rancher instead of RKE etc?

Thanks!

**Comment by Yurgil** [comment]
Still waiting this feature

**Comment by LH-IG** [comment]
+++ for this feature
